svm.thy

```diff
@@ -0,0 +1,353 @@
+
+
+lemma asd :
+  assumes "fst h \<in> H"
+  assumes "is_arg_min (soft_svm) (\<lambda> x . fst x\<in>H \<and> (\<forall>i\<in>{0..<n}. snd (S i) * (inner (fst x) (fst (S i))) \<ge> 1 - (snd x i)) \<and> 
+    (\<forall> i\<in>{0..<n}. snd x i \<ge> 0) ) h"
+  shows "case h of (s,ksi) \<Rightarrow> \<forall>i\<in>{0..<n}. ksi i = hinge_loss s (S i)"
+proof 
+
+  fix s ksi
+  assume "h = (s,ksi)"
+  let ?P = "(\<lambda> x . fst x\<in>H \<and> (\<forall>i\<in>{0..<n}. snd (S i) * (inner (fst x) (fst (S i))) \<ge> 1 - (snd x i)) \<and> 
+    (\<forall> i\<in>{0..<n}. snd x i \<ge> 0) )"
+  have "s\<in>H" using `fst h \<in> H` `h = (s,ksi)` by auto  
+  have 2:"(\<forall>i\<in>{0..<n}. snd (S i) * (inner s (fst (S i))) \<ge> 1 - (ksi i))"
+    using  `h = (s,ksi)` `is_arg_min (soft_svm) ?P h` 
+    by (simp add: is_arg_min_linorder)
+  have 3:" (\<forall> i\<in>{0..<n}. ksi i \<ge> 0)"
+    using  `h = (s,ksi)` `is_arg_min (soft_svm) ?P h` 
+    by (simp add: is_arg_min_linorder)
+
+
+  let ?hinge = "(\<lambda> i. hinge_loss s (S i))"
+  have "soft_svm (s,ksi) \<le> soft_svm (s, ?hinge)" 
+  proof -
+    have 1: "is_arg_min (soft_svm) ?P (s,ksi)"
+      using `h = (s,ksi)`  `is_arg_min (soft_svm) ?P h`  by auto
+    have " ?P (s, ?hinge)" unfolding hinge_loss.simps   using `s\<in>H`  by auto
+    then show ?thesis using 1 
+      by (simp add: is_arg_min_linorder)
+  qed
+  then have "(1/n)* (\<Sum>i\<in>{0..<n}. (ksi i)) \<le> 
+            (1/n)* (\<Sum>i\<in>{0..<n}. (?hinge i))"  by (simp add: local.soft_svm_def)
+  then have 4:"(\<Sum>i\<in>{0..<n}. (ksi i)) \<le>  (\<Sum>i\<in>{0..<n}. (?hinge i))" 
+    by (smt divide_le_0_1_iff n_pos not_one_le_zero of_nat_le_0_iff real_mult_le_cancel_iff2)
+  have 5:"\<forall> i \<in> {0..<n}. ksi i \<ge> ?hinge i" unfolding hinge_loss.simps using 2 3 by auto
+  then have "(\<Sum>i\<in>{0..<n}. (ksi i)) \<ge>  (\<Sum>i\<in>{0..<n}. (?hinge i))"
+    by (meson sum_mono)
+  then have 6:"(\<Sum>i\<in>{0..<n}. (ksi i)) =  (\<Sum>i\<in>{0..<n}. (?hinge i))" using 4 by auto
+
+
+  show "\<forall>i\<in>{0..<n}.  ksi i = hinge_loss s (S i)"
+  proof 
+    fix i
+    assume "i\<in>{0..<n}" 
+    show " ksi i = hinge_loss s (S i)" 
+    proof (rule ccontr)
+      assume "ksi i \<noteq> hinge_loss s (S i)" 
+      then have "ksi i > ?hinge i" using 5 `i\<in>{0..<n}`
+        by fastforce
+      then have "(\<Sum>i\<in>{0..<n}. (ksi i)) >  (\<Sum>i\<in>{0..<n}. (?hinge i))" 
+        by (meson "5" \<open>i \<in> {0..<n}\<close> finite_atLeastLessThan sum_strict_mono_ex1)
+      then show False using 6 by auto
+    qed
+  qed
+qed
+
+lemma asd1 :
+  assumes " h \<in> H"
+  assumes "is_arg_min (soft_svm) (\<lambda> x . fst x\<in>H \<and> 
+  (\<forall>i\<in>{0..<n}. snd (S i) * (inner (fst x) (fst (S i))) \<ge> 1 - (snd x i)) \<and> 
+    (\<forall> i\<in>{0..<n}. snd x i \<ge> 0) ) (h, ksi)"
+  shows "\<forall>i\<in>{0..<n}. ksi i = hinge_loss h (S i)" using asd 
+  using assms(1) assms(2) by auto
+
+lemma 
```

Comment:
this is claim 15.5? the proof needs to be refactored, it is hard to follow! how does it relate to the informal proof in the book?